src/test/java/org/sonar/squidbridge/ProgressAstScannerTest.java

```diff
@@ -63,4 +73,82 @@ public boolean matches(final Object argument) {
     }));
   }
 
+  @Test(timeout = 5000)
+  public void testInterrupt() throws Exception {
+    SquidAstVisitorContextImpl<Grammar> context = new SquidAstVisitorContextImpl<Grammar>(new SourceProject(""));
+    CountDownLatch latch = new CountDownLatch(1);
+    ProgressReport progress = mock(ProgressReport.class);
+    Parser<Grammar> parser = new BlockingParser(latch);
+
+    AstScanner<Grammar> scanner = new ProgressAstScanner.Builder<Grammar>(context)
+      .setProgressReport(progress)
+      .setBaseParser(parser)
+      .setFilesMetric(MiniCMetrics.FILES)
+      .build();
+
+    File[] files = {new File("src/test/resources/metrics/lines.mc"), new File("src/test/resources/metrics/lines2.mc")};
+    ScannerRunner runner = new ScannerRunner(scanner, files);
+    runner.start();
+
+    // wait for first parsing to start
+    latch.await();
+    verify(progress).start(any(Collection.class));
+
+    runner.interrupt();
+    Awaitility.waitAtMost(Duration.TWO_SECONDS).until(new AssertProgressReportCancelled(progress));
+    verifyNoMoreInteractions(progress);
+  }
+
+  private class AssertProgressReportCancelled implements Runnable {
+    private final ProgressReport progress;
+
+    AssertProgressReportCancelled(ProgressReport progress) {
+      this.progress = progress;
+    }
+
+    public void run() {
+      verify(progress).cancel();
+    }
+  }
+
+  private static class BlockingParser extends Parser<Grammar> {
+    private final CountDownLatch latch;
+
+    protected BlockingParser(CountDownLatch latch) {
+      super(null);
+      this.latch = latch;
+    }
+
+    @Override
+    public AstNode parse(File f) {
+      latch.countDown();
+      try {
+        Thread.sleep(10_000);
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Remove this use of "Thread.sleep()". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2925)